### PR TITLE
Bids 1591/broadcast voluntary exits

### PR DESCRIPTION
**Pull request by guybrush**

### CODE REVIEW

File: db/node_jobs.go [hunk 1]
```diff
@@ -193,111 +243,172 @@ func UpdateBLSToExecutionChangesNodeJob(job *types.BLSToExecutionChangesNodeJob)
...
 	if err != nil {
 		return err
 	}
 	if resp.StatusCode != 200 {
 		d, _ := ioutil.ReadAll(resp.Body)
-		return fmt.Errorf("http request error: %s: %s, data: %s", resp.Status, d, data)
+		return fmt.Errorf("http request error: %s: %s, data: %s", resp.Status, d, job.RawData)
 	}
-	_, err = WriterDb.Exec(`update node_jobs set status = $1 where id = $2`, types.SubmittedToNodeNodeJobStatus, job.ID)
+	job.Status = types.SubmittedToNodeNodeJobStatus
```
Comment (D13ce) [inline db/node_jobs.go:274]: **Minor:** This is duplicated code, see line 406. You should move it into a function, something like `submitJobInDb`.

File: db/node_jobs.go [hunk 2]
```diff
@@ -66,28 +124,19 @@ func SubmitNodeJobs() error {
...
-	}
+	nj.ID = uuid.New().String()
+	nj.Status = types.PendingNodeJobStatus
 
 	opsByIndex := map[uint64]*capella.SignedBLSToExecutionChange{}
 	opsToCheck := map[uint64]bool{}
 	indicesArr := []uint64{}
-	for _, op := range nj.Data {
-		err = verifyBlsToExecutionChangeSignature(op)
+	d, _ := nj.GetBLSToExecutionChangesNodeJobData()
```
Comment (D13ce) [inline db/node_jobs.go:137]: Why not check for the error here?

File: types/node_jobs.go [hunk 3]
```diff
@@ -20,58 +21,88 @@ type NodeJobType string
...
+			if nj.Type != "" && nj.Type != UnknownNodeJobType && nj.Type != VoluntaryExitsNodeJobType {
+				return fmt.Errorf("nodejob.RawData missmatches nodejob.Type (%v)", nj.Type)
+			}
+			nj.Type = VoluntaryExitsNodeJobType
+			nj.Data = d
+			return nj.SanitizeRawData()
+		}
+	}
+	nj.Type = UnknownNodeJobType
+	return fmt.Errorf("can not unmarshal job-data")
```
Comment (D13ce) [inline types/node_jobs.go:88]: **Minor:** This should be changed to "cannot". Furthermore I suggest to capitalize the first letter as this is shown as flash messages on the webpage.

File: types/node_jobs.go [hunk 4]
```diff
@@ -20,58 +21,88 @@ type NodeJobType string
...
+type NodeJobValidatorInfo struct {
+	ValidatorIndex      uint64 `db:"validatorindex"`
+	PublicKey           []byte `db:"pubkey"`
+	WithdrawCredentials []byte `db:"withdrawalcredentials"`
+	ExitEpoch           uint64 `db:"exitepoch"`
+	Status              string
+}
+
+// ParseData will try to unmarshal NodeJob.RawData into NodeJob.Data and determine NodeJob.Type by doing so. If it is not able to unmarshal any type it will return an error. It will sanitize NodeJob.RawData on success.
+func (nj *NodeJob) ParseData() error {
```
Comment (D13ce) [inline types/node_jobs.go:61]: **Minor:** I suggest to check for empty data first (when the user presses the submit button without adding json data) and if the data is empty, show a different error message than the one in line 88.

**Review by D13ce — COMMENTED** [review]
Overall: lgtm

I have found some minor issues and added a few questions. Feel free to simply resolve anything labeled **Minor**.

Furthermore:

**Minor:** The following functions should start with a lower-case letter:
- CreateBLSToExecutionChangesNodeJob
- UpdateBLSToExecutionChangesNodeJob
- SubmitBLSToExecutionChangesNodeJobs
- SubmitBLSToExecutionChangesNodeJob
- CreateVoluntaryExitNodeJob
- UpdateVoluntaryExitNodeJobs
- UpdateVoluntaryExitNodeJob
- SubmitVoluntaryExitNodeJobs
- SubmitVoluntaryExitNodeJob

File: db/node_jobs.go [hunk 5]
```diff
@@ -2,30 +2,29 @@ package db
...
-	"github.com/pkg/errors"
-	ssz "github.com/prysmaticlabs/go-ssz"
 	"github.com/sirupsen/logrus"
-	e2types "github.com/wealdtech/go-eth2-types/v2"
 	ethutil "github.com/wealdtech/go-eth2-util"
 )
 
-func init() {
-	err := e2types.InitBLS()
+func GetCurrentForkVersion() (*types.ForkVersion, error) {
```
[the remark this answers is not in the document]
Comment (guybrush) [reply db/node_jobs.go]: fixed

File: db/node_jobs.go [hunk 6]
```diff
@@ -193,111 +243,172 @@ func UpdateBLSToExecutionChangesNodeJob(job *types.BLSToExecutionChangesNodeJob)
...
+		return err
+	}
+	if strings.HasPrefix(dbResult.Status, "exit") {
+		job.Status = types.CompletedNodeJobStatus
+		job.CompletedTime.Time = time.Now()
+		job.CompletedTime.Valid = true
+		_, err = WriterDb.Exec(`update node_jobs set status = $1, completed_time = $2 where id = $3`, job.Status, job.CompletedTime.Time, job.ID)
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (guybrush) [reply db/node_jobs.go:363]: duplicate but less complex this way

File: db/node_jobs.go (hunk 2 above)
Comment (guybrush) [reply db/node_jobs.go]: fixed

File: types/node_jobs.go (hunk 4 above)
Comment (guybrush) [reply types/node_jobs.go:61]: updated

**Review by D13ce — APPROVED** [review]
lgtm

**Pull request merged**